File: androidApp/src/main/java/com/nordbones/weatherapp/android/screens/home/HomeScreen.kt
package com.nordbones.weatherapp.android.screens.home

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.res.stringResource
import androidx.navigation.NavController
import com.google.accompanist.swiperefresh.SwipeRefresh
import com.google.accompanist.swiperefresh.rememberSwipeRefreshState
import com.nordbones.weatherapp.android.R
import com.nordbones.weatherapp.android.navigation.NavigationTree
import com.nordbones.weatherapp.android.screens.home.widgets.Card
import com.nordbones.weatherapp.android.screens.home.widgets.MainWeatherWidget
import com.nordbones.weatherapp.android.utils.Handle
import com.nordbones.weatherapp.domain.model.Weather
import com.nordbones.weatherapp.viewmodel.ViewState


@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun HomeScreen(viewModel: HomeViewModel, navController: NavController) {
    val weatherState by viewModel.weatherState.collectAsState(initial = ViewState(isLoading = true))

    LaunchedEffect(key1 = Unit) {
        viewModel.getWeather()
    }

    Scaffold(
        modifier = Modifier.fillMaxSize(),
        floatingActionButton = {
            ExtendedFloatingActionButton(onClick = { navController.navigate(route = NavigationTree.SETTINGS.navTarget) }) {
                Icon(imageVector = Icons.Default.Settings, contentDescription = null)
                Text(text = stringResource(R.string.settings))
            }
        }) { paddingValues ->
        SwipeRefresh(
            state = rememberSwipeRefreshState(isRefreshing = false),
            onRefresh = { viewModel.getWeather() }) {
            HomeScreenContent(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(
                        bottom = paddingValues.calculateBottomPadding(),
                        top = paddingValues.calculateTopPadding()
                    ),
                state = weatherState,
            )
        }
    }
}


@Composable
fun HomeScreenContent(
    modifier: Modifier,
    state: ViewState<Weather>,
) {
    val scrollState = rememberScrollState()
    state.Handle { weather ->
        Column(
            modifier = modifier
                .padding(dimensionResource(id = R.dimen.dp20))
                .verticalScroll(scrollState),
            verticalArrangement = Arrangement.Center
        ) {
            MainWeatherWidget(weather = weather)

            Row {
                Card(
                    modifier = Modifier.weight(1f),
                    label = stringResource(R.string.pressure),
                    value = weather.airPressure.toString(),
                    units = stringResource(R.string.millibars)
                )
                Card(
                    modifier = Modifier.weight(1f),
                    label = stringResource(R.string.humidity),
                    value = weather.humidityInPercents.toString(),
                    units = stringResource(R.string.percents)
                )
            }
            Row {
                Card(
                    modifier = Modifier.weight(1f),
                    label = stringResource(R.string.visibility),
                    value = weather.visibilityInMiles.toString(),
                    units = stringResource(R.string.miles)
                )
                Card(
                    modifier = Modifier.weight(1f),
                    label = stringResource(R.string.wind),
                    value = weather.windSpeedInMpH.toString(),
                    units = stringResource(R.string.mph)
                )
            }
        }
    }
}

File: iosApp/iosApp/Screens/Home/HomeScreen.swift
//
//  HomeScreen.swift
//  iosApp
//
//  Created by Никита on 27.05.22.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import shared


struct HomeScreen: View {
    
    @ObservedObject var viewModel = HomeViewModel()
    
    private var state: ViewState<Weather> {
        get{
            return viewModel.state
        }
    }
    
    var body: some View {
        VStack {
            HomeScreenContent(state: state,onRefresh: {viewModel.getWeather()})
        }.toolbar {
            NavigationLink(destination: SettingsScreen()){
                Label("Settings", systemImage: "gear")
            }
        }.onAppear {
            viewModel.getWeather()
        }
    }
}


struct HomeScreenContent : View {
    private  let refreshControlKey = "RefreshControl"
    
    let state : ViewState<Weather>
    let onRefresh : () -> Void
    
    var body: some View {
        ScrollView {
            RefreshControl(coordinateSpace: .named(refreshControlKey)) {
                onRefresh()
            }
            ViewStateHandler(state: state) { data in
                VStack(alignment: .center) {
                    MainWeatherWidget(weather: data)
                    
                    HStack {
                        Card(label: "Pressure", value: String(format: "%.2f", data.airPressure), units: "milibars")
                        Card(label: "Humidity", value: String(format: "%.2f", data.humidityInPercents), units: "%")
                    }
                    HStack {
                        Card(label: "Visibilit", value: String(format: "%.2f", data.visibilityInMiles), units: "miles")
                        Card(label: "Wind", value: String(format: "%.2f", data.windSpeedInMpH), units: "mph")
                    }
                }.padding()
                
            } error: { errorMessage in
                Text("Error: \(errorMessage)")
            } loading: {
                ProgressView()
            }
            
        }.coordinateSpace(name: refreshControlKey)
    }
}
